diffusion/model.py

```diff
@@ -0,0 +1,119 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from utils import (
+    cosine_beta_schedule,
+    default,
+    extract,
+    unnormalize_to_zero_to_one,
+)
+
+class DiffusionModel(nn.Module):
+    def __init__(
+        self,
+        model,
+        timesteps = 1000,
+        sampling_timesteps = None,
+        ddim_sampling_eta = 1.
+    ):
+        super(DiffusionModel, self).__init__()
+        assert model.channels == model.out_dim
+        assert not model.learned_sinusoidal_cond
+
+        self.model = model
+        self.channels = self.model.channels
+        self.self_condition = self.model.self_condition
+        self.device = torch.cuda.current_device()
+
+        self.betas = cosine_beta_schedule(timesteps).to(self.device)
+        self.num_timesteps = self.betas.shape[0]
+
+        alphas = 1. - self.betas
+        self.alphas_cumprod = torch.cumprod(alphas, dim=0)
+        self.alphas_cumprod_prev = F.pad(self.alphas_cumprod[:-1], (1, 0), value = 1.)
+    
+        # calculations for diffusion q(x_t | x_{t-1}) and others
+        self.sqrt_alphas_cumprod = torch.sqrt(self.alphas_cumprod)
+        self.sqrt_one_minus_alphas_cumprod = torch.sqrt(1. - self.alphas_cumprod)
+        self.log_one_minus_alphas_cumprod =  torch.log(1. - self.alphas_cumprod)
+        self.sqrt_recip_alphas_cumprod = torch.sqrt(1. / self.alphas_cumprod)
+        self.sqrt_recipm1_alphas_cumprod = torch.sqrt(1. / self.alphas_cumprod - 1)
+
+        # calculations for posterior q(x_{t-1} | x_t, x_0)
+        self.posterior_variance = self.betas * (
+            1. - self.alphas_cumprod_prev) / (1. - self.alphas_cumprod)
+
+        self.posterior_log_variance_clipped = torch.log(
+            self.posterior_variance.clamp(min =1e-20))
+        self.posterior_mean_coef1 = self.betas * torch.sqrt(
+            self.alphas_cumprod_prev) / (1. - self.alphas_cumprod)
+        self.posterior_mean_coef2 = (1. - self.alphas_cumprod_prev) * torch.sqrt(alphas) / (
+            1. - self.alphas_cumprod)
```

Comment:
we need to blank these, these are all part of the equations

I would try to keep some of the variable names, and add helper functions which we can ask them to fill out.

The only ones I would keep here are:
`alphas, alphas_cumprod, posterior_variance, posterior_mean_coef1 (comment that this is first part of mu in the README), posterior_mean_coef2 (comment that this is the second part of mu in the README)`

For example:

`self.posterior_variance = self._get_posterior_variance(...)`
`self.posterior_log_variance_clipped = torch.log(self.posterior_variance...)`

The rest of the fields, they can create and name however they want. Only those 3 are dependent on the clamping / normalization things.